### PR TITLE
feat: Add getAll methods for header, body and path params

### PR DESCRIPTION
**Pull request by ebebbington**
Addresses #382 

**Description**

* Adds method to get all header params

* Adds method to get all body params

**Other Notes**
* Decided against the get all path params, as i can't figure out a solution atm, should we still merge this PR? it only provides 2 out of the 3 methods it was supposed to

### CODE REVIEW

**Comment by MVEMCJSUNPE** [comment]
I'm not sure if getAllPathParams was in the issue... I remember that 'getAllBodyParams', 'getAllQueryParams', 'getAllHeaderParams', and 'all' methods were mentioned. But if getAllPathParams was in the issue, can't we just return the path_params property in class Request?

**Comment by ebebbington** [comment]
`path_params` is an empty object at that point - the server class creates the path params **using the resource**, but nice suggestion, that's what i tried first :P

File: src/http/request.ts [hunk 1]
```diff
@@ -147,6 +148,30 @@ export class Request extends ServerRequest {
...
+    for (const pair of this.headers.entries()) {
+      headers[pair[0]] = pair[1];
+    }
+    return headers;
+  }
+
+  /**
+   * Gets all the body params
+   *
+   * @return The parsed body as an object
```
Comment (crookse) [inline src/http/request.ts:167]: @returns

File: src/http/request.ts [hunk 2]
```diff
@@ -147,6 +148,30 @@ export class Request extends ServerRequest {
     return this.headers.get(input);
   }
 
+  /**
+   * Gets all header params
+   *
+   * @return Key value pairs for the header name and it's value
```
Comment (crookse) [inline src/http/request.ts:154]: @returns

Comment (crookse) [inline src/http/request.ts:154]:
```suggestion
   * @returns Key value pairs for the header name and it's value
```

File: src/http/request.ts (hunk 1 above)
Comment (crookse) [reply src/http/request.ts:167]:
```suggestion
   * @returns The parsed body as an object
```

File: tests/unit/http/request_test.ts [hunk 3]
```diff
@@ -399,6 +407,75 @@ function getBodyParamTests() {
...
+    "Returns all of the path parameters",
+    async () => {
+      const serverRequest = members.mockRequest();
+      const request = new Drash.Http.Request(serverRequest);
+      await request.parseBody();
+      request.path_params = {
+        hello: "world",
+        goodbye: "world",
+      };
+      //const actual = request.getAllPathParams();
```
Comment (crookse) [inline tests/unit/http/request_test.ts:470]: no assertions?

File: src/http/request.ts (hunk 2 above)
Comment (ebebbington) [reply src/http/request.ts:154]: seems @returns isn't something the docs support, but @return is: https://doc.deno.land/https/deno.land/x/sockets@v0.5.0/mod.ts

Comment (crookse) [reply src/http/request.ts:154]: wow that's garbage.... tsdoc states `@returns` -_-. alright then

File: tests/unit/http/request_test.ts (hunk 3 above)
Comment (ebebbington) [reply tests/unit/http/request_test.ts:470]: oh right yeah, thats cause i started adding the tests but didnt go with the feature for path params, ill comment the test out

Comment (ebebbington) [reply tests/unit/http/request_test.ts]: https://github.com/drashland/deno-drash/pull/404/commits/9db5da3879ecb02604853a57b64d7bc676b05165

**Comment by crookse** [comment]
im going to try to tackle the path params stuff

**Comment by ebebbington** [comment]
Cheers Eric!

**Review by crookse — APPROVED** [review]
LGTM BROTHER

**Comment by MVEMCJSUNPE** [comment]
Maybe it's just me being stupid, but I thought I suggested to @ebebbington that getAllPathParams could just return this.path_params in one of the issue comments above (this [one](https://github.com/drashland/deno-drash/pull/404#issuecomment-716866638), I believe). Then @ebebbington rebuked that it wouldn't work because "`path_params `is an empty object at that point - the server class creates the path params **using the resource**" in this [issue comment](https://github.com/drashland/deno-drash/pull/404#issuecomment-716868175).

**Comment by ebebbington** [comment]
Turns out, I made a mistake lol :p A very dumb moment, I wasn’t ‘hydrating ‘ the request object, which does happen during the lifecycle of the request

**Comment by ebebbington** [comment]
It’s not you being stupid, it’s me :p

**Comment by MVEMCJSUNPE** [comment]
Nah, I'm not even sure if I understand the drash life cycle or what REST is (lol). From the sound of it, it seems like there's actually a lot of complicated things going on that I'm not aware of.

**Comment by ebebbington** [comment]
Feel free to take a look at the code that starts the whole request lifecycle, it’s the handleHttpRequest method I believe in src/http/server.ts :)

A bare basic web server would just listen for requests on a host name and port, and be able to handle that request

**Comment by crookse** [comment]
> fyi this is in staging, will close once it's in prod

this needs to be merged in before we deploy staging docs to production

**Comment by ebebbington** [comment]
im getting the issue and pr mixed up lol

**Pull request merged**